### PR TITLE
Report onscreen text to Fuchsia context engine.

### PR DESCRIPTION
**Pull request by tbmbob**
Add a class to track onscreen text via SemanticsNodes from the
accessibility layer, and report this text to the ContextEngine.

### CODE REVIEW

**Comment by Hixie** [comment]
cc @goderbauer

File: content_handler/accessibility_bridge.cc [hunk 1]
```diff
@@ -0,0 +1,79 @@
+// Copyright 2017 The Fuchsia Authors. All rights reserved.
+// Use of this source code is governed by a BSD-style license that can be
+// found in the LICENSE file.
+
+#include "flutter/content_handler/accessibility_bridge.h"
+
+#include <unordered_set>
+
+#include "apps/maxwell/services/context/context_writer.fidl.h"
+#include "apps/modular/lib/rapidjson/rapidjson.h"
```
Comment (abarth) [inline content_handler/accessibility_bridge.cc]: I'd recommend just using rapidjson directory.  These wrappers don't really add much (and aren't worth taking the dependency).

File: content_handler/accessibility_bridge.h [hunk 2]
```diff
@@ -0,0 +1,42 @@
...
+#include "flutter/lib/ui/semantics/semantics_node.h"
+#include "lib/app/cpp/application_context.h"
+
+namespace flutter_runner {
+
+// Maintain an up-to-date list of SemanticsNodes on screen, and communicate
+// with the Context Service.
+class AccessibilityBridge {
+ public:
+  AccessibilityBridge(app::ApplicationContext* context);
```
Comment (abarth) [inline content_handler/accessibility_bridge.h]: explicit

File: content_handler/runtime_holder.cc [hunk 3]
```diff
@@ -269,6 +273,8 @@ void RuntimeHolder::CreateView(
 
   runtime_->dart_controller()->dart_state()->SetReturnCodeCallback(
       [this](int32_t return_code) { return_code_ = return_code; });
+
+  runtime_->SetSemanticsEnabled(true);
```
Comment (abarth) [inline content_handler/runtime_holder.cc]: Are we sure we want to do this unconditionally?  Enabling semantics will have an impact on performance.

**Comment by abarth** [comment]
Generally LGTM.  We discussed the following in person:

1. @goderbauer should review the logic for maintaining the state in the accessibility bridge.
2. Hold off enabling semantics by default until we understand the performance trade-off.
3. After this patch lands, we should follow up with the Maxwell team to see if we can come up with a lower bandwidth protocol for sending updates.  (Currently, the protocol requires re-sending all the text on the page even for only one word has changed.)

File: content_handler/accessibility_bridge.cc [hunk 4]
```diff
@@ -0,0 +1,79 @@
...
+void AccessibilityBridge::UpdateSemantics(
+    const std::vector<blink::SemanticsNode>& update) {
+  for (const auto& node : update) {
+    semantics_nodes_[node.id] = node;
+  }
+  std::vector<int> visited_nodes;
+  UpdateVisitedForNodeAndChildren(0, &visited_nodes);
+  EraseUnvisitedNodes(visited_nodes);
+
+  // The data sent to the Context Service is a JSON formatted list of strings
```
Comment (goderbauer) [inline content_handler/accessibility_bridge.cc]: What is contained in `label` is not necessary shown on screen. For example, an `IconButton` to open a menu might have the label "Open menu", which is not necessarily visible on screen.

**Comment by goderbauer** [comment]
> @goderbauer should review the logic for maintaining the state in the accessibility bridge.

LGTM

File: content_handler/accessibility_bridge.cc (hunk 1 above)
Comment (tbmbob) [reply content_handler/accessibility_bridge.cc]: Done!

File: content_handler/accessibility_bridge.h (hunk 2 above)
Comment (tbmbob) [reply content_handler/accessibility_bridge.h]: Done.

File: content_handler/runtime_holder.cc (hunk 3 above)
Comment (tbmbob) [reply content_handler/runtime_holder.cc]: I've removed this until we have more data.

File: content_handler/accessibility_bridge.cc (hunk 4 above)
Comment (tbmbob) [reply content_handler/accessibility_bridge.cc]: I updated the comment to be more explicit about this. Is it fair to say that the collection of labels is a superset of the on screen text?

Comment (Hixie) [reply content_handler/accessibility_bridge.cc]: Not necessarily. For example, there might be decorative text on the screen that isn't included in the accessibility data.

**Comment by tbmbob** [comment]
The new version of my commit just made changes in response to your comments. You guys have to merge the pull, correct? I summarize next plans after this CL below.

Short term:

- Test performance & enable accessibility and/or add a toggle option of some sort.
- Better ContextWriting. Instead of reporting all text on screen together, each widget / render object can have its own Value.

Long term:

- Instead of passing full strings, Context Engine will hold a handle to an Entity managed by Flutter Engine.

**Comment by tbmbob** [comment]
cc @gabeschine

File: content_handler/BUILD.gn [hunk 5]
```diff
@@ -48,6 +50,8 @@ template("flutter_content_handler") {
              "//garnet/public/lib/ui/scenic:client",
              "//garnet/public/lib/ui/input/fidl",
              "//garnet/public/lib/ui/views/fidl",
+             "//apps/maxwell/services/context",
+             "//apps/modular/lib/rapidjson:rapidjson",
```
Comment (abarth) [inline content_handler/BUILD.gn]: Please use `//third_party/rapidjson` directly.

File: content_handler/accessibility_bridge.cc [hunk 6]
```diff
@@ -0,0 +1,84 @@
...
+#include "third_party/rapidjson/rapidjson/stringbuffer.h"
+#include "third_party/rapidjson/rapidjson/writer.h"
+
+namespace flutter_runner {
+
+AccessibilityBridge::AccessibilityBridge(app::ApplicationContext* context)
+    : writer_(context->ConnectToEnvironmentService<maxwell::ContextWriter>()) {
+  writer_.set_connection_error_handler(
+      [] { FXL_LOG(ERROR) << "Error connecting to ContextWriter"; });
+  FXL_VLOG(2) << "Constructing AccessibilityBridge.";
```
Comment (abarth) [inline content_handler/accessibility_bridge.cc]: Please remove this log message.  This object is created every time.

File: content_handler/accessibility_bridge.h [hunk 7]
```diff
@@ -0,0 +1,42 @@
...
+  // Walk the semantics node tree starting at |id|, and store the id of each
+  // visited child in |visited_nodes|.
+  void UpdateVisitedForNodeAndChildren(const int id,
+                                       std::vector<int>* visited_nodes);
+
+  // Remove any node from |semantics_nodes_| that doesn't have an id in
+  // |visited_nodes|.
+  void EraseUnvisitedNodes(const std::vector<int>& visited_nodes);
+
+  std::unordered_map<int, blink::SemanticsNode> semantics_nodes_;
```
Comment (abarth) [inline content_handler/accessibility_bridge.h]: You're probably better off using `std::map` here.  `std::map` is almost always what you want due to sadness in the spec for `std::unordered_map`.

File: content_handler/runtime_holder.cc [hunk 8]
```diff
@@ -160,6 +161,9 @@ void RuntimeHolder::Init(
     blink::SetRegisterNativeServiceProtocolExtensionHook(
         ServiceProtocolHooks::RegisterHooks);
   }
+
+  accessibility_bridge_ = std::unique_ptr<AccessibilityBridge>(
+      new AccessibilityBridge(context_.get()));
```
Comment (abarth) [inline content_handler/runtime_holder.cc]: I think we slightly prefer `std::make_unique`

**Comment by abarth** [comment]
LGTM

... with some tiny nits.  Happy to merge once the nits are cleared up.

**Comment by tbmbob** [comment]
Thanks! Suggestions fixed.

**Comment by abarth** [comment]
Thanks!

**Pull request merged**